File: lib/msu/command.rb
module MSU
  class Command
    def initialize(displays)
      @primary, *@connected = displays.connected
      @disconnected = displays.disconnected
      @xrandr_arguments = ['xrandr']
    end

    def extend_right
      connected.reduce(primary) do |last, display|
        off(display).on(display).right_of(last)

        display
      end

      self
    end

    def all_off
      disconnected.each { |display| off(display) }

      self
    end

    def off(display)
      add "--output #{display.name} --off"
    end

    def on(display)
      add "--output #{display.name}"
      add "--mode #{display.modes.first}"
    end

    def right_of(display)
      add "--right-of #{display.name}"
    end

    def to_s
      xrandr_arguments.join(' ')
    end

    private

    attr_reader :primary, :connected, :disconnected, :xrandr_arguments

    def add(command)
      xrandr_arguments << command

      self
    end
  end
end

File: MSU/Xrandr/Command.hs
module MSU.Xrandr.Command where

import Control.Monad
import Control.Monad.Writer
import MSU.Display

type Xrandr a = Writer String a

output :: Display -> Xrandr ()
output d = tell $ " --output " ++ name d

off :: Xrandr ()
off = tell " --off"

mode :: Mode -> Xrandr ()
mode m = tell $ " --mode " ++ show m

rightOf :: Display -> Xrandr ()
rightOf d = tell $ " --right-of " ++ name d

outputOff :: Display -> Xrandr ()
outputOff d = output d >> off

-- | Return @False@ if the output was not connected.
outputOn :: Display -> Xrandr Bool
outputOn   (Display _ []   ) = return False
outputOn d@(Display _ (m:_)) = output d >> mode m >> return True

allOff :: [Display] -> Xrandr ()
allOff = mapM_ outputOff

firstOn :: [Display] -> Xrandr ()
firstOn []    = return ()
firstOn (d:_) = outputOn d >> return ()

extendRight :: [Display] -> Xrandr ()
extendRight = fold1M_ (placeWith rightOf)

placeWith :: (Display -> Xrandr ()) -> Display -> Display -> Xrandr Display
placeWith placementCmd primary secondary = do
    outputOff secondary
    outputOn secondary &&> placementCmd primary

    return secondary

buildCommand :: (Xrandr a) -> String
buildCommand f = execWriter $ tell "xrandr" >> f

-- | Executes the second action IFF the first returns @True@.
(&&>) :: Monad m => m Bool -> m () -> m ()
f &&> g = f >>= \b -> if b then g else return ()

-- | Like foldM_ but uses first element as base value
fold1M_ :: Monad m => (b -> b -> m b) -> [b] -> m ()
fold1M_ _ [] = return ()
fold1M_ f (x:xs) = foldM_ f x xs
